r2-shared/Publication/PublicationManifest.ts

```diff
@@ -0,0 +1,196 @@
+    this.resources = manifestJSON.resources ? new Links(manifestJSON.resources) : new Links([]);
+    this.tableOfContents = manifestJSON.toc ? new Links(manifestJSON.toc) : new Links([]);
+
+    this.manifestUrl = manifestUrl;
+  }
+
+  // Getting Manifest
+
+  public static async getManifest(manifestUrl: URL, store?: Store): Promise<PublicationManifest> {
+    const fetchManifest = async (): Promise<PublicationManifest> => {
+      const response = await window.fetch(manifestUrl.href, {
+        credentials: "same-origin"
+      })
+      const manifestJSON = await response.json();
+      if (store) {
+        await store.set("manifest", JSON.stringify(manifestJSON));
+      }
+      return new PublicationManifest(manifestJSON, manifestUrl);
+    };
+
+    const tryToUpdateManifestButIgnoreResult = async (): Promise<void> => {
+      try {
+        await fetchManifest();
+      } catch (err) {
+        // Ignore errors.
+      }
+      return new Promise<void>(resolve => resolve());
+    }
+
+    // Respond immediately with the manifest from the store, if possible.
+    if (store) {
+      const manifestString = await store.get("manifest");
+      if (manifestString) {
+        // Kick off a fetch to update the store for next time,
+        // but don't await it.
+        tryToUpdateManifestButIgnoreResult();
+        const manifestJSON = JSON.parse(manifestString);
+        return new PublicationManifest(manifestJSON, manifestUrl);
+      }
+    }
+
+    return fetchManifest();
+  }
+
+  // Getting readingOrder items
+
+  public getCoverLink(): Link | null {
+    return this.linkWithRel("cover");
+  }
+
+  public getStartLink(): Link | null {
+    if (this.readingOrder.length > 0) {
+      return this.readingOrder[0];
+    }
+    return null;
+  }
+
+  public getPreviousSpineItem(href: string): Link | null {
+    const index = this.getSpineIndex(href);
+    if (index !== null && index > 0) {
+      return this.readingOrder[index - 1];
+    }
+    return null;
+  }
+
+  public getNextSpineItem(href: string): Link | null {
+    const index = this.getSpineIndex(href);
+    if (index !== null && index < (this.readingOrder.length - 1)) {
+      return this.readingOrder[index + 1];
+    }
+    return null;
+  }
+
+  public getSpineItem(href: string): Link | null {
+    const index = this.getSpineIndex(href);
+    if (index !== null) {
+      return this.readingOrder[index];
+    }
+    return null;
+  }
+
+  public getSpineIndex(href: string): number | null {
+    for (let index = 0; index < this.readingOrder.length; index++) {
+      const item = this.readingOrder[index];
+      if (item.href) {
+        const itemUrl = new URL(item.href, this.manifestUrl.href).href;
+        if (itemUrl === href) {
+          return index;
+        }
+      }
+    }
+    return null;
+  }
+
+  public getAbsoluteHref(href: string): string | null {
+    return new URL(href, this.manifestUrl.href).href;
+  }
```

Comment:
The preferred way to get the absolute HREF is through `link.toURL(publication.baseURL())`